### PR TITLE
[fix] create redis vector index before adding docs, add prefix to doc…

### LINKED ISSUE #11197: Documents not being correctly indexed in vector database. ["redis.exceptions.ResponseError: my_docs: no such index"]
### System Info

Following the steps of indexing from [ https://python.langchain.com/docs/modules/data_connection/indexing ](url) you'll find the following error "redis.exceptions.ResponseError: my_docs: no such index".
You'll get this exception while using redis as retriever:
Screenshot from 2023-09-28 16-57-03

### Who can help?

_No response_

### Information

- [x] The official example notebooks/scripts
- [ ] My own modified scripts

### Related Components

- [ ] LLMs/Chat Models
- [ ] Embedding Models
- [ ] Prompts / Prompt Templates / Prompt Selectors
- [ ] Output Parsers
- [ ] Document Loaders
- [X] Vector Stores / Retrievers
- [ ] Memory
- [ ] Agents / Agent Executors
- [ ] Tools / Toolkits
- [ ] Chains
- [ ] Callbacks/Tracing
- [ ] Async

### Reproduction

Screenshot from 2023-09-28 16-13-18
The error is here:
Screenshot from 2023-09-28 16-32-00
If you look for the index you'll get (empty list or set).
This line makes impossible to save in the wanted format, and there's another thing: The index is not created for some reason.
I'll try to fix, but I'm not sure if it's possible for me at the moment, so I'm reporting this, I hope it helps.

### Expected behavior

Expected behavior inside Redis: "docs:indexname_:12ss2sadd"

### PR DESCRIPTION
**Pull request by a9raag**
Fix Description: 
For Redis Vector integration in add_texts method, there were two issues that lead to this bug. 
1. Vector index is not being created leading to no such_index error 
2. `doc:index` prefix was also missing for Redis Keys. 

resolves #11197 
Maintainer: @baskaryan

### CODE REVIEW

**Comment by a9raag** [comment]
@hwchase17 @eyurtsev 
Can someone take a look at this? I am getting 
```
google-github-actions/auth failed with: retry function failed after 4 attempts: the GitHub Action workflow must specify exactly one of "workload_identity_provider" or "credentials_json"! If you are specifying input values via GitHub secrets, ensure the secret is being injected into the environment. By default, secrets are not passed to workflows triggered from forks, including Dependabot.
```

File: libs/langchain/langchain/vectorstores/redis/base.py [hunk 1]
```diff
@@ -707,9 +707,14 @@ def add_texts(
 
         # Write data to redis
         pipeline = self.client.pipeline(transaction=False)
+        if not embeddings:
+            embeddings = []
+            for i, text in enumerate(texts):
```
Comment (Spartee) [inline libs/langchain/langchain/vectorstores/redis/base.py:712]: The embeddings are made a few lines down? if we did create the index here, why loop through them all?

File: libs/langchain/langchain/vectorstores/redis/base.py [hunk 2]
```diff
@@ -707,9 +707,14 @@ def add_texts(
...
+                embeddings.append(self._embeddings.embed_query(text))
+        self._create_index(len(embeddings[0]))
```
Comment (Spartee) [inline libs/langchain/langchain/vectorstores/redis/base.py:714]: The workflow is technically to use ``from_documents`` and then ``add_texts`` but this is the second time I've heard this feedback (https://github.com/langchain-ai/langchain/issues/9585#issuecomment-1716729646) so I'm starting to think people just use ``add_texts``.

**Review by Spartee — COMMENTED** [review]
Ok so I see the point about the prefix, but there are also people that want to define the entire key themselves. This is because they retrieve from the redis store from outside langchain as well and would like to be able to have transparent key names.

This was a design decision made here: https://github.com/langchain-ai/langchain/issues/10411 

I feel like there is an argument to be made for both. 

On the point of re-initializing the index in this function, this has also come up. Should an index **always** be created an any method ``add_*`` or ``from_*`` if it does not exist? I'm happy to make that change.

This is the workflow I know most people use with Redis


The workflow is 
1. use one of the class methods ex. ``from_documents``
2. use it

and then to add to existing
1. use ``from_existing``
2. use ``add_documents``/ ``add_texts`` as necessary
3. use it.

@baskaryan feedback??

**Review by baskaryan — COMMENTED** [review]
for backwards compatibility + optionality, updated to check if prefix is already in the key and only add if it isn't. what do we think?

**Comment by tylerhutcherson** [comment]
> for backwards compatibility + optionality, updated to check if prefix is already in the key and only add if it isn't. what do we think?

@Spartee @baskaryan LGTM >> My only input is that the current implementation, and this adjustment, still don't provide flexibility to allow for users to define their own index prefix right? It uses `_redis_prefix()` to generate this everytime. This will eventually need to change, but just highlighting that for future improvements.

**Comment by a9raag** [comment]
@tylerhutcherson 
The current implementation does provide users the ability to define their own index prefix. 
While creating the Redis vector store object, user can provide the `index_name` argument, (see [base.py#L251](https://github.com/a9raag/langchain/blob/059a0c54ab9350ce93871ddb727ebbd879630c43/libs/langchain/langchain/vectorstores/redis/base.py#L251)) which is then used to generate the prefix for the document. `_redis_prefix(self.index_name)`  (see [base.py#L684](https://github.com/a9raag/langchain/blob/059a0c54ab9350ce93871ddb727ebbd879630c43/libs/langchain/langchain/vectorstores/redis/base.py#L684))
For example, 
```py
vector_store = Redis(
        redis_url = "localhost:6379", 
        index_name = "custom_prefix", 
        embedding = embedding)
```
will generate documents with key prefix `doc:custom_prefix:1234`

**Comment by Spartee** [comment]
@a9raag, what Tyler is talking about is the “doc:” portion. You’re code shows the custom index name. 

Users may want the flexibility to have multiple indicies with different prefixes, yet similar keys. (Or some combination of this). The previous methodology seemed introduced as a convenience, but limited the functionality. 

Maybe we add the ability to define prefix? Seems like it might be a footgun.

**Comment by tylerhutcherson** [comment]
> @a9raag, what Tyler is talking about is the “doc:” portion. You’re code shows the custom index name.
> 
> Users may want the flexibility to have multiple indicies with different prefixes, yet similar keys. (Or some combination of this). The previous methodology seemed introduced as a convenience, but limited the functionality.
> 
> Maybe we add the ability to define prefix? Seems like it might be a footgun.

^^ exactly -- redis is meant to be multi-tenant by design. So requiring the key prefix with `doc:` up front is an anti-pattern for redis. This should be a configurable setting, but could come in a later revision for the sake of simplicity. happy to help :)

File: libs/langchain/langchain/vectorstores/redis/base.py [hunk 3]
```diff
@@ -667,6 +653,10 @@ def drop_index(
             # Index not exist
             return False
 
+    @property
+    def _full_key_prefix(self) -> str:
```
Comment (baskaryan) [inline libs/langchain/langchain/vectorstores/redis/base.py:657]: @tylerhutcherson @Spartee is this what you had in mind?

**Review by Spartee — APPROVED** [review]
This looks good to me. A couple comments though

This assumes that people are going to use the ``prefix:index_name`` convention if they pass their own keys despite us only requiring to pass the ``prefix`` portion. Since I think this could be confusing, we should at least explain this in the initializer and then I think we should have a simple property method like

```python

def key_prefix(self) -> str:
     return self._full_key_prefix() + ":"
```



also, a comment in the class level or the method level of anything that creates an index with something like the following would be good

```text
This method will create an index in Redis if one does not exist. 

keys for the records will be automatically generated.

Keys can be passed in and used manually. If the keys are not prefixed, they will be prefixed with the output from the ``Redis.key_prefix()`` method which is a combination of the prefix passed in the initializer and the index name.
```

Overall I think being clear on the behavior here is paramount.

**Pull request merged**